sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala

```diff
@@ -0,0 +1,367 @@
+ *                           be compressed individually.
+ */
+abstract class BatchedTableCompressor(maxBatchMemorySize: Long) extends AutoCloseable with Arm {
+  private[this] val tables = new ArrayBuffer[ContiguousTable]
+  private[this] val tempBuffers = new ArrayBuffer[DeviceMemoryBuffer]
+  private[this] val oversizedOutBuffers = new ArrayBuffer[DeviceMemoryBuffer]
+  private[this] val results = new ArrayBuffer[CompressedTable]
+
+  // temporary and output memory being used as part of the current batch
+  private[this] var batchMemUsed: Long = 0
+
+  /**
+   * Add a contiguous table to be batch-compressed. Ownership of the table is transferred to the
+   * batch compressor which is responsible for closing the table.
+   * @param contigTable the contiguous table to be compressed
+   */
+  def addTable(contigTable: ContiguousTable): Unit = {
+    closeOnExcept(contigTable) { contigTable =>
+      val tempSize = getTempSpaceNeeded(contigTable.getBuffer)
+      var memNeededThisBuffer = tempSize
+      if (batchMemUsed + memNeededThisBuffer > maxBatchMemorySize) {
+        compressBatch()
+      }
+      val tempBuffer = if (tempSize > 0) {
+        DeviceMemoryBuffer.allocate(memNeededThisBuffer)
+      } else {
+        null
+      }
+      closeOnExcept(tempBuffer) { tempBuffer =>
+        val outputSize = getOutputSpaceNeeded(contigTable.getBuffer, tempBuffer)
+        memNeededThisBuffer += outputSize
+        if (batchMemUsed + memNeededThisBuffer > maxBatchMemorySize) {
+          compressBatch()
+        }
+        oversizedOutBuffers += DeviceMemoryBuffer.allocate(outputSize)
+        tempBuffers += tempBuffer
+        tables += contigTable
+        batchMemUsed += memNeededThisBuffer
+      }
+    }
+  }
+
+  /**
+   * Add an array of contiguous tables to be compressed. The tables will be closed by the
+   * batch compressor.
+   * @param contigTable contiguous tables to compress
+   */
+  def addTables(contigTable: Array[ContiguousTable]): Unit = {
+    var i = 0
+    try {
+      contigTable.foreach { ct =>
+        addTable(ct)
+        i += 1
+      }
+    } catch {
+      case t: Throwable =>
+        contigTable.drop(i).foreach(_.safeClose())
+        throw t
+    }
+  }
+
+  /**
+   * This must be called after all tables to be compressed have been added to retrieve the
+   * compression results.
+   * @note the table IDs in the TableMeta of all tables will be set to zero
+   * @return compressed tables
+   */
+  def finish(): Array[CompressedTable] = {
+    // compress the last batch
+    compressBatch()
+
+    val compressedTables = results.toArray
+    results.clear()
+    compressedTables
+  }
+
+  /** Must be closed to release the resources owned by the batch compressor */
+  override def close(): Unit = {
+    tables.safeClose()
+    tempBuffers.safeClose()
+    oversizedOutBuffers.safeClose()
+    results.safeClose()
+  }
+
+  private def compressBatch(): Unit = {
+    if (tables.nonEmpty) {
```

Comment:
nit, could move this if to the `compressBatch` method definition, so spacing goes to the left.